[BE] Job feedback

MihaiDruta opened the pull request:
[BE] Job feedback
https://trello.com/c/tSobOP3D/86-be-job-feedback

1.Created Feedback entity
2.Added possibility to give feedback to users, 1-5 stars + comment
(owner -> user involved, user involved -> owner, user involved -> user involved / only if activity is finished)
3.Added possibility to edit your feedback given to someone
4.The average score of an user is shown in his profile as stars (1-5)
5.Added possibility to see someone's full feedback history (paginated + sorted by creation time or stars)
6.Fully working documentation with all necessary models and query parameters

Commit:
[BE] Job feedback
https://trello.com/c/tSobOP3D/86-be-job-feedback

Commit:
[BE] Job feedback
https://trello.com/c/tSobOP3D/86-be-job-feedback

1.Created Feedback entity
2.Added possibility to give feedback to users, 1-5 stars + comment
(owner -> user involved, user involved -> owner, user involved -> user involved / only if activity is finished)
3.Added possibility to edit your feedback given to someone
4.The average score of an user is shown in his profile as stars (1-5)
5.Added possibility to see someone's full feedback history (paginated + sorted by creation time or stars)
6.Fully working documentation with all necessary models and query parameters

SpartakusMd commented on src/Controller/FeedbackController.php:
```diff
@@ -0,0 +1,438 @@
+     * ),
+     *     @SWG\Parameter(
+     *     description="Json body for the request",
+     *     name="requestBody",
```
Did you mean comment?

SpartakusMd commented on src/Controller/FeedbackController.php:
```diff
@@ -0,0 +1,438 @@
+     *     type="integer",
+     * ),
+     *     @SWG\Parameter(
+     *     description="Json body for the request",
```
I think you forgot to update the description. It should be the message fior feedback.

SpartakusMd commented on src/Controller/FeedbackController.php:
```diff
@@ -0,0 +1,438 @@
+            $userTo->setStars($newFeedback->getStars());
+        } else {
+            $countFeedback = $this->feedbackRepository->countFeedback($userTo)->getQuery()->getSingleScalarResult();
+            $newStars = (($userTo->getStars() * ($countFeedback - 1)) + $newFeedback->getStars()) / $countFeedback;
```
1. getQuery() and getSingleScalarResult() should always be in a repository.
2. What's up with this hard math logic? You will slowly over time get a wrong value because of estimation. The correct way is to have FeedbackRepository calculate this for you via a query. Something like `SELECT sum(stars)/count(id) as stars from feedback where user_id = ?`
3. If 2 will be implemented you won't have this `if-else`

Don't forget about the editFeedback()

SpartakusMd commented on src/Entity/User.php:
```diff
@@ -66,7 +73,7 @@ class User implements UserInterface
      * @ORM\Column(type="string", unique=true)
      * @Serializer\Expose()
-     * @Groups({"UserDetail", "UserList", "ActivityList"})
+     * @Groups({"UserDetail", "UserList", "ActivityList", "FeedbackList"})
```
Maybe it makes sense to have `name` and `surname` also available in these groups?

SpartakusMd commented on src/Filters/FeedbackSort.php:
```diff
@@ -0,0 +1,36 @@
+     */
+    public $stars;
+
+    private function isSortingParamValid($sortingParameter): bool
```
This method is used in every `*Sort` class. Extract it to some base abstract class.

Commit:
[BE] Job feedback
https://trello.com/c/tSobOP3D/86-be-job-feedback

1.Changed some parameters description for documentation
2.Added a function to FeedbackRepository to retrieve the sum of stars for an User
3.With the help of the new function in FeedbackRepository now the logic of calculating current amount of stars is much more simple
4.Added name and surname to FeedbackList and ActivityList groups
5.Created BaseSorting class which contains validating function for sorting parameters in query

SpartakusMd commented on src/Controller/FeedbackController.php:
```diff
@@ -0,0 +1,429 @@
+        $newFeedback = $this->feedbackTransformer->addFeedback($feedbackDTO, $authenticatedUser, $activity, $userTo);
+        $this->feedbackRepository->save($newFeedback);
+
+        $userTo->setStars($this->feedbackRepository->getStarsSum($userTo)
```
I don't see any other usages for `getStarsSum()` and `countFeedback()`. Why not create a single fnction which returns the final result?

Commit:
[BE] Job feedback
https://trello.com/c/tSobOP3D/86-be-job-feedback

1.Removed getStarsSum() and countFeedback() methods
2.Added getAvgStars() instead which does the same
